ddl/table.go

```diff
@@ -152,6 +138,74 @@ func onCreateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 	}
 }
 
+func getBundleFromDBInfo(t *meta.Meta, job *model.Job, dbInfo *model.DBInfo) (*placement.Bundle, error) {
+	dbBundleWrapper := func(bundle *placement.Bundle) *placement.Bundle {
+		bundle.Index = placement.RuleIndexDatabase
+		bundle.Override = true
+		return bundle
+	}
+	if dbInfo.DirectPlacementOpts != nil {
+		// check the direct placement option compatibility.
+		if err := checkPolicyValidation(dbInfo.DirectPlacementOpts); err != nil {
+			return nil, errors.Trace(err)
+		}
+		// build bundle from direct placement options.
+		bundle, err := placement.NewBundleFromOptions(dbInfo.DirectPlacementOpts)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return dbBundleWrapper(bundle), nil
+	}
+	if dbInfo.PlacementPolicyRef != nil {
+		// placement policy reference will override the direct placement options.
+		po, err := checkPlacementPolicyExistAndCancelNonExistJob(t, job, dbInfo.PlacementPolicyRef.ID)
+		if err != nil {
+			return nil, errors.Trace(infoschema.ErrPlacementPolicyNotExists.GenWithStackByArgs(dbInfo.PlacementPolicyRef.Name))
+		}
+		// build bundle from placement policy.
+		bundle, err := placement.NewBundleFromOptions(po.PlacementSettings)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return dbBundleWrapper(bundle), nil
+	}
+	return nil, nil
+}
+
+func getBundleFromTblInfo(t *meta.Meta, job *model.Job, tbInfo *model.TableInfo, dbInfo *model.DBInfo) (*placement.Bundle, error) {
+	tableBundleWrapper := func(bundle *placement.Bundle) *placement.Bundle {
```

Comment:
We can avoid this by checking the bundle at end of the `getBundleFromTblInfo` like this:

```
if bundle != nil {
  // do some bundle modification.
}
```

Just a advices.